// @vscode/vscode-perf/out/auth.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateVscodeDevAuthState = generateVscodeDevAuthState;
const node_fetch_1 = __importDefault(require("node-fetch"));
const cookie = __importStar(require("cookie"));
const js_base64_1 = __importDefault(require("js-base64"));
const crypto_1 = require("crypto");
function* flatten(...iterables) {
    for (const iterable of iterables) {
        for (const el of iterable) {
            yield el;
        }
    }
}
async function encrypt(serverKey, data) {
    const iv = crypto_1.webcrypto.getRandomValues(new Uint8Array(12));
    const clientKeyObj = await crypto_1.webcrypto.subtle.generateKey({ name: 'AES-GCM', length: 256 }, true, ['encrypt', 'decrypt']);
    const clientKey = new Uint8Array(await crypto_1.webcrypto.subtle.exportKey('raw', clientKeyObj));
    const keyData = new Uint8Array(32);
    for (let i = 0; i < keyData.byteLength; i++) {
        keyData[i] = clientKey[i] ^ serverKey[i];
    }
    const key = await crypto_1.webcrypto.subtle.importKey('raw', keyData, { name: 'AES-GCM', length: 256 }, true, ['encrypt', 'decrypt']);
    const dataBase64 = js_base64_1.default.encode(data);
    const dataUint8Array = js_base64_1.default.toUint8Array(dataBase64);
    const cipherText = await crypto_1.webcrypto.subtle.encrypt({ name: 'AES-GCM', iv, }, key, dataUint8Array);
    const result = new Uint8Array(flatten(clientKey, iv, new Uint8Array(cipherText)));
    return js_base64_1.default.fromUint8Array(result);
}
/**
 * Generates a Playwright storage state file for the  given GitHub username and token.
 *
 * Make sure the token has the ['repo', 'workflow', 'user:email', 'read:user'] scopes.
 */
async function generateVscodeDevAuthState(token) {
    const authResponse = await (0, node_fetch_1.default)('https://auth.insiders.vscode.dev/', { method: 'POST' });
    if (!authResponse.ok) {
        throw new Error(`Failed to reach auth endpoint: ${authResponse.status} ${authResponse.statusText}`);
    }
    const setCookieHeaders = authResponse.headers.raw()['set-cookie'];
    const vscodeSessionHeader = setCookieHeaders.find(header => header.startsWith('vscode.session='));
    const vscodeSessionCookie = cookie.parse(vscodeSessionHeader);
    const rawServerKey = await authResponse.text();
    const serverKey = js_base64_1.default.toUint8Array(rawServerKey);
    if (serverKey.byteLength != 32) {
        throw Error('The key retrieved by the server is not 32 bytes long.');
    }
    const credentials = [{
            account: 'github.auth',
            password: JSON.stringify({
                extensionId: 'vscode.github-authentication',
                content: JSON.stringify([{
                        accessToken: token,
                        account: { label: 'GitHub', id: 0 },
                        id: 'github',
                        scopes: ['repo', 'workflow', 'user:email', 'read:user']
                    }])
            }),
            service: 'vscode-insidersvscode.github-authentication'
        }];
    return {
        cookies: [
            {
                name: "vscode.session",
                value: vscodeSessionCookie['vscode.session'],
                domain: "auth.insiders.vscode.dev",
                path: "/",
                expires: new Date(vscodeSessionCookie['expires']).getTime() / 1000,
                httpOnly: true,
                secure: false,
                sameSite: "Strict"
            }
        ],
        origins: [
            {
                origin: "https://insiders.vscode.dev",
                localStorage: [
                    {
                        name: "credentials.provider",
                        value: await encrypt(serverKey, JSON.stringify(credentials))
                    }
                ]
            }
        ]
    };
}
